### PR TITLE
bugfix/STJS-607-visa-doesnt-work-on-develop

### PR DESCRIPTION
**Pull request by pgs-gmartowski**

### CODE REVIEW

File: src/application/core/integrations/ApplePay.spec.ts [hunk 1]
```diff
@@ -1078,9 +1082,16 @@ function ApplePayFixture() {
...
     requestTypes: ['AUTH']
   };
+  const configProvider = mock(ConfigProvider);
+  const communicator = mock(InterFrameCommunicator);
   const jwt =
     'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJsaXZlMl9hdXRvand0IiwiaWF0IjoxNTUzMjcwODAwLCJwYXlsb2FkIjp7ImJhc2VhbW91bnQiOiIxMDAwIiwiY3VycmVuY3lpc28zYSI6IkdCUCIsInNpdGVyZWZlcmVuY2UiOiJsaXZlMiIsImFjY291bnR0eXBlZGVzY3JpcHRpb24iOiJFQ09NIn19.SGLwyTcqh6JGlrgzEabOLvCWRx_jeroYk67f_xSQpLM';
-  const instance = new ApplePay(config, jwt, 'https://example.com');
+
+  when(communicator.whenReceive(anyString())).thenReturn({
+    thenRespond: () => undefined
```
Comment (wirus15) [inline src/application/core/integrations/ApplePay.spec.ts:1091]: `thenRespond` should return rxjs stream, so replace `undefined` with e.g. `EMPTY`

File: src/application/core/integrations/ApplePay.ts [hunk 2]
```diff
@@ -121,25 +125,39 @@ export class ApplePay {
   private _merchantId: string;
   private _paymentRequest: any;
   private _placement: string;
-  private readonly _completion: { errors: []; status: string };
+  private _completion: { errors: []; status: string };
   private _localStorage: BrowserLocalStorage;
+  private readonly _config$: Observable<IConfig>;
+  private _config: IWalletConfig;
```
Comment (wirus15) [inline src/application/core/integrations/ApplePay.ts:131]: you could rename this to `applePayConfig` because now it seems like it can be confused with `_config$`

File: src/application/core/integrations/ApplePay.ts [hunk 3]
```diff
@@ -121,25 +125,39 @@ export class ApplePay {
...
-    this.jwt = jwt;
-    this._completion = {
-      errors: [],
-      status: ApplePaySession ? this.getPaymentSuccessStatus() : ''
-    };
-    this._configurePaymentProcess(jwt, config, gatewayUrl);
-    this._onInit(config.buttonText, config.buttonStyle);
+
+    this._config$ = this._configProvider.getConfig$();
+    this._communicator.whenReceive(MessageBus.EVENTS_PUBLIC.CONFIG_CHECK).thenRespond(() => this._config$);
```
Comment (wirus15) [inline src/application/core/integrations/ApplePay.ts:140]: This can be removed. It was for config synchronization between ST and ControlFrame (I'm not sure if it's still being used).

File: src/application/core/integrations/ApplePay.ts [hunk 4]
```diff
@@ -121,25 +125,39 @@ export class ApplePay {
...
+
+    this._config$.subscribe(config => {
+      const { applePay, jwt, datacenterurl } = config;
+      if (applePay) {
+        this._config = applePay;
+      }
+      this._localStorage.setItem('completePayment', '');
```
Comment (wirus15) [inline src/application/core/integrations/ApplePay.ts:147]: those properties that don't come from the `config$` can be set up outside the `subscribe()`

File: src/application/core/integrations/ApplePayMock.spec.ts [hunk 5]
```diff
@@ -131,8 +135,15 @@ function applePayMockFixture() {
     buttonText: 'donate',
     buttonStyle: 'white-outline'
   };
+  const configProvider = mock(ConfigProvider);
+  const communicator = mock(InterFrameCommunicator);
+  when(communicator.whenReceive(anyString())).thenReturn({
+    thenRespond: () => undefined
```
Comment (wirus15) [inline src/application/core/integrations/ApplePayMock.spec.ts:141]: thenRespond: () => EMPTY

File: src/application/core/integrations/VisaCheckoutMock.spec.ts [hunk 6]
```diff
@@ -1,18 +1,26 @@
...
+  const jwt =
+    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJsaXZlMl9hdXRvand0IiwiaWF0IjoxNTUzMjcwODAwLCJwYXlsb2FkIjp7ImJhc2VhbW91bnQiOiIxMDAwIiwiY3VycmVuY3lpc28zYSI6IkdCUCIsInNpdGVyZWZlcmVuY2UiOiJsaXZlMiIsImFjY291bnR0eXBlZGVzY3JpcHRpb24iOiJFQ09NIn19.SGLwyTcqh6JGlrgzEabOLvCWRx_jeroYk67f_xSQpLM';
   // when
   beforeEach(() => {
-    const { config, livestatus } = VisaCheckoutMockFixture();
-    const jwt =
-      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJsaXZlMl9hdXRvand0IiwiaWF0IjoxNTUzMjcwODAwLCJwYXlsb2FkIjp7ImJhc2VhbW91bnQiOiIxMDAwIiwiY3VycmVuY3lpc28zYSI6IkdCUCIsInNpdGVyZWZlcmVuY2UiOiJsaXZlMiIsImFjY291bnR0eXBlZGVzY3JpcHRpb24iOiJFQ09NIn19.SGLwyTcqh6JGlrgzEabOLvCWRx_jeroYk67f_xSQpLM';
-    instance = new VisaCheckoutMock(config, jwt, 'https://example.com', livestatus);
+    when(communicator.whenReceive(anyString())).thenReturn({
+      thenRespond: () => undefined
```
Comment (wirus15) [inline src/application/core/integrations/VisaCheckoutMock.spec.ts:20]: EMPTY

File: src/application/core/integrations/VisaCheckout.spec.ts [hunk 7]
```diff
@@ -7,12 +11,39 @@ jest.mock('../../../../src/application/core/shared/Notification');
...
+  const jwt =
+    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJsaXZlMl9hdXRvand0IiwiaWF0IjoxNTUzMjcwODAwLCJwYXlsb2FkIjp7ImJhc2VhbW91bnQiOiIxMDAwIiwiY3VycmVuY3lpc28zYSI6IkdCUCIsInNpdGVyZWZlcmVuY2UiOiJsaXZlMiIsImFjY291bnR0eXBlZGVzY3JpcHRpb24iOiJFQ09NIn19.SGLwyTcqh6JGlrgzEabOLvCWRx_jeroYk67f_xSQpLM';
   // when
   beforeEach(() => {
-    const { config, livestatus, url } = VisaCheckoutFixture();
-    const jwt =
-      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJsaXZlMl9hdXRvand0IiwiaWF0IjoxNTUzMjcwODAwLCJwYXlsb2FkIjp7ImJhc2VhbW91bnQiOiIxMDAwIiwiY3VycmVuY3lpc28zYSI6IkdCUCIsInNpdGVyZWZlcmVuY2UiOiJsaXZlMiIsImFjY291bnR0eXBlZGVzY3JpcHRpb24iOiJFQ09NIn19.SGLwyTcqh6JGlrgzEabOLvCWRx_jeroYk67f_xSQpLM';
-    instance = new VisaCheckout(config, jwt, url, livestatus);
+    when(communicator.whenReceive(anyString())).thenReturn({
+      thenRespond: () => undefined
```
Comment (wirus15) [inline src/application/core/integrations/VisaCheckout.spec.ts:21]: EMPTY

Comment (wirus15) [reply src/application/core/integrations/VisaCheckout.spec.ts:21]: btw, if this is form `CONFIG_CHECK` then it can be removed too

File: src/application/core/integrations/VisaCheckout.ts [hunk 8]
```diff
@@ -78,7 +82,10 @@ export class VisaCheckout {
   private _notification: NotificationService;
   private _stJwt: StJwt;
   private _livestatus: number = 0;
+  private _datacenterurl: string;
   private _placement: string = 'body';
+  private readonly _config$: Observable<IConfig>;
+  private _config: IWalletConfig;
```
Comment (wirus15) [inline src/application/core/integrations/VisaCheckout.ts:88]: this can be renamed to `_visaCheckoutConfig`

File: src/application/core/integrations/VisaCheckout.ts [hunk 9]
```diff
@@ -90,17 +97,26 @@ export class VisaCheckout {
...
+  constructor(private _configProvider: ConfigProvider, private _communicator: InterFrameCommunicator) {
     this._messageBus = Container.get(MessageBus);
     this._notification = Container.get(NotificationService);
-    config.requestTypes = config.requestTypes !== undefined ? config.requestTypes : ['AUTH'];
-    this._stJwt = new StJwt(jwt);
-    this._livestatus = livestatus;
-    this._configurePaymentProcess(jwt, config, gatewayUrl);
-    this._initVisaFlow();
+    this._config$ = this._configProvider.getConfig$();
+    this._communicator.whenReceive(MessageBus.EVENTS_PUBLIC.CONFIG_CHECK).thenRespond(() => this._config$);
```
Comment (wirus15) [inline src/application/core/integrations/VisaCheckout.ts:104]: this to be removed

File: src/application/core/integrations/ApplePay.spec.ts (hunk 1 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/ApplePay.spec.ts]: done

File: src/application/core/integrations/ApplePay.ts (hunk 2 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/ApplePay.ts]: done

File: src/application/core/integrations/ApplePay.ts (hunk 3 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/ApplePay.ts]: ok, done

Comment (pgs-gmartowski) [reply src/application/core/integrations/ApplePay.ts]: it's used in `ST.ts`

File: src/application/core/integrations/ApplePay.ts (hunk 4 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/ApplePay.ts]: done, I only left `onInit()` in ApplePay inside - it looks like it was triggered before config was set when I'd moved it outside of subscribe.

File: src/application/core/integrations/ApplePayMock.spec.ts (hunk 5 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/ApplePayMock.spec.ts]: done

File: src/application/core/integrations/VisaCheckoutMock.spec.ts (hunk 6 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/VisaCheckoutMock.spec.ts]: done

File: src/application/core/integrations/VisaCheckout.spec.ts (hunk 7 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/VisaCheckout.spec.ts]: done

File: src/application/core/integrations/VisaCheckout.ts (hunk 8 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/VisaCheckout.ts]: done

File: src/application/core/integrations/VisaCheckout.ts (hunk 9 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/VisaCheckout.ts]: done

File: src/application/core/integrations/ApplePay.ts (hunk 3 above)
Comment (pgs-gmartowski) [reply src/application/core/integrations/ApplePay.ts]: and in `ControlFrame.ts` as well; without it iframes are not displayed

File: src/client/config/ConfigResolver.ts [hunk 10]
```diff
@@ -48,7 +48,7 @@ export class ConfigResolver {
     return {
       analytics: this._getValueOrDefault(config.analytics, false),
       animatedCard: this._getValueOrDefault(config.animatedCard, false),
-      applePay: this._setApmConfig(config.applePay, config.components),
+      applePay: this._setApmConfig(config.applePay, config.applePay),
```
Comment (wirus15) [inline src/client/config/ConfigResolver.ts:51]: why this change? If it was intended then I think one `config.applePay` would be enough :)

Comment (pgs-gmartowski) [reply src/client/config/ConfigResolver.ts:51]: 🤦 here should be a default version of applePay, but it's in the other branch, sorry for that ;)

Comment (pgs-gmartowski) [reply src/client/config/ConfigResolver.ts:51]: reverted
